classes/key_manager.php

```diff
@@ -0,0 +1,144 @@
+<?php
+// This file is part of Moodle - http://moodle.org/
+//
+// Moodle is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// Moodle is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with Moodle.  If not, see <http://www.gnu.org/licenses/>.
+
+/**
+ * Access key manager.
+ *
+ * @package    tool_pdfpages
+ * @author     Tom Dickman <tomdickman@catalyst-au.net>
+ * @copyright  2021 Catalyst IT
+ * @license    http://www.gnu.org/copyleft/gpl.html GNU GPL v3 or later
+ */
+
+namespace tool_pdfpages;
+
+defined('MOODLE_INTERNAL') || die();
+
+/**
+ * Access key manager.
+ *
+ * @package    tool_pdfpages
+ * @author     Tom Dickman <tomdickman@catalyst-au.net>
+ * @copyright  2021 Catalyst IT
+ * @license    http://www.gnu.org/copyleft/gpl.html GNU GPL v3 or later
+ */
+class key_manager {
+
+    /**
+     * Create a user key.
+     *
+     * @param int $instanceid the instance ID to create key for.
+     * @param string $iprestriction optional IP range to restrict access to.
+     *
+     * @return string the created user key value.
+     * @throws \moodle_exception if user doesn't have permission to create key.
+     */
+    public static function create_user_key(int $instanceid, string $iprestriction = ''): string {
+        global $USER;
+
+        require_capability('tool/pdfpages:generatepdf', \context_system::instance());
+
+        $iprestriction = !empty($iprestriction) ? $iprestriction : null;
+
+        self::delete_user_key($instanceid);
+
+        $ttl = get_config('tool_pdfpages', 'accesskeyttl');
+        $expirationtime = !empty($ttl) ? (time() + $ttl) : (time() + MINSECS);
+
+        return create_user_key('tool/pdfpages', $USER->id, $instanceid, $iprestriction, $expirationtime);
+    }
+
+    /**
+     * Delete a user key.
+     *
+     * @param int $instanceid the instance ID to delete user key for.
+     *
+     * @return bool true on success.
+     */
+    public static function delete_user_key(int $instanceid): bool {
```

Comment:
should it be delete_keys as it maybe potentially more than one?  Would be probably good to add userid as a parameter if we want to use it for a specific user.

Reply:
Renames and inject user ID rather than relying on global